### PR TITLE
Relax Socket.EnableBroadcast getter exception behavior

### LINKED ISSUE #70555: Reading the property "EnableBroadcast" on a "System.Net.Sockets.Socket" throws an exception and closes the socket
### Description

TCP Socket is closed when `Socket.EnableBroadcast` is read.

### Reproduction Steps

1. Enter this code into `main()`:
```
var  tcpClient  = new TcpClient ("127.0.0.1", 54321);
bool connected1 = tcpClient.Connected;

try { _ = tcpClient.Client.EnableBroadcast; }
catch (Exception) {}
bool connected2 = tcpClient.Connected;
```

2. Open a TCP Server terminal software like "HW Group Hercules", start a TCP Server on port e.g. 54321.
3. Run the code.
4. The value of `connected1` will be true, but the value of `connected2` will be false.

This behaviour is more or less correct, since the documentation of `Socket.EnableBroadcast` (https://docs.microsoft.com/en-us/dotnet/api/system.net.sockets.socket.enablebroadcast?view=net-6.0) actually says that it will throw an exception if the Socket is _not_ a datagram socket (which means, if it is not UDP):

> [SocketException]
> This option is valid for a datagram socket only.

The problem arises if the properties of the socket are queried in the DataTip or Locals Window or Autos Window: In this case, the `EnableBroadcast` property *IS* read, which causes the unwanted closing of the socket. This is quite annoying because it means that DataTip on TCP Sockets is not usable.

First time showing the DataTip:
grafik
Second time showing the DataTip:
grafik

### Expected behavior

Socket must not be closed. -> `EnableBroadcast` must not be shown in DataTip and Locals Window and others, or it must return false if not supported. The `set` of `EnableBroadcast` is not called in the DataTip, so this one may still throw.

### Actual behavior

Socket is closed due to the exception that happens when showing the Socket properties in the DataTip.

### Regression?

_No response_

### Known Workarounds

_No response_

### Configuration

.NET 5
Win 10 Enterprise 1809 x64
Architecture: x64
Most likely it's unrelated to the .NET version, OS and architecture.

### Other information

I guess that there are similar bugs in other classes, where properties must not be read in certain circumstances, but they will be read by the DataTip.

### PR DESCRIPTION
**Pull request by turnyanszki**
Fixes #70555

### CODE REVIEW

**Comment by dnfadmin** [comment]
[CLA assistant check](https://cla.dotnetfoundation.org/dotnet/runtime?pullRequest=70728) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla.dotnetfoundation.org/dotnet/runtime?pullRequest=70728) before we can accept your contribution.<br/><br/>:x: turnyanszki [sign now](https://cla.dotnetfoundation.org/dotnet/runtime?pullRequest=70728)<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla.dotnetfoundation.org/check/dotnet/runtime?pullRequest=70728) it.</sub>

File: src/libraries/System.Net.Sockets/tests/FunctionalTests/System.Net.Sockets.Tests.csproj [hunk 1]
```diff
@@ -55,50 +55,31 @@
     <Compile Include="SocketTaskExtensionsTest.cs" />
     <Compile Include="MulticastOptionTest.cs" />
     <Compile Include="UdpClientTest.cs" />
+    <Compile Include="EnableBroadcastSocketTest.cs" />
     <Compile Include="UnixDomainSocketTest.cs" />
     <!-- Common Sockets files -->
-    <Compile Include="$(CommonTestPath)System\Net\Configuration.cs"
```
Comment (antonfirsov) [inline src/libraries/System.Net.Sockets/tests/FunctionalTests/System.Net.Sockets.Tests.csproj:60]: Can we revert the unrelated formatting changes? I know it's Visual Studio autocorrecting our practices, but still :)

**Review by antonfirsov — COMMENTED** [review]
Thanks for the contribution! A few comments on tests, otherwise good. Would be nice if someone else could take a look at this to make 100% sure it's the right thing to change this, @wfurt maybe?

File: src/libraries/System.Net.Sockets/tests/FunctionalTests/EnableBroadcastSocketTest.cs [hunk 2]
```diff
@@ -0,0 +1,47 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using Xunit;
+
+namespace System.Net.Sockets.Tests
+{
+    public class EnableBroadcastSocketTest
```
[the remark this answers is not in the document]
Comment (turnyanszki) [reply src/libraries/System.Net.Sockets/tests/FunctionalTests/EnableBroadcastSocketTest.cs:8]: I thougt this naming could be useful in the future if someone wants to add more test to the broadcast functionality, but of course I can rename that.

File: src/libraries/System.Net.Sockets/tests/FunctionalTests/System.Net.Sockets.Tests.csproj (hunk 1 above)
Comment (turnyanszki) [reply src/libraries/System.Net.Sockets/tests/FunctionalTests/System.Net.Sockets.Tests.csproj:60]: Yes, sorry.

File: src/libraries/System.Net.Sockets/tests/FunctionalTests/EnableBroadcastSocketTest.cs (hunk 2 above)
[the remark this answers is not in the document]
Comment (antonfirsov) [reply src/libraries/System.Net.Sockets/tests/FunctionalTests/EnableBroadcastSocketTest.cs]: Then we can rename it at that point :) It's a thin wrapper around a native socket feature, I see no need to test it that extensively.

**Comment by turnyanszki** [comment]
@antonfirsov  Could you help me please why these tests are failing? They work for me at local.

**Comment by rzikm** [comment]
> How can i run the other OS tests at local?

Unfortunately, you can't "just run" these tests on other OSes, you need either a VM or WSL installation with that os and recompile the entire runtime in there.

**Comment by turnyanszki** [comment]
> > How can i run the other OS tests at local?
> 
> Unfortunately, you can't "just run" these tests on other OSes, you need either a VM or WSL installation with that os and recompile the entire runtime in there.
Yeah I tought :( Any idea or tip why these are failing? Should I make the test windows specific or debug them in all OS instances?

**Comment by rzikm** [comment]
I did a quick experiment with the original repro and it seems that it does not reproduce on Linux:

```cs
using System.Net.Sockets;
using System.Net;

var tcpListener = new TcpListener(54321);
tcpListener.Start();

var tcpClient = new TcpClient("127.0.0.1", 54321);
bool connected1 = tcpClient.Connected;
System.Console.WriteLine($"Connected1: {connected1}");

try { _ = tcpClient.Client.EnableBroadcast; }
catch (Exception) { }
bool connected2 = tcpClient.Connected;
System.Console.WriteLine($"Connected2: {connected2}");
```

```
 radekzikmund@radekzikmund-desktop  C:  source  DotnetSandbox                                                                                                                                                                                        10:06:24  
❯ dotnet run 
C:\source\DotnetSandbox\Program.cs(4,19): warning CS0618: 'TcpListener.TcpListener(int)' is obsolete: 'This constructor has been deprecated. Use TcpListener(IPAddress localaddr, int port) instead.' [C:\source\DotnetSandbox\DotnetSandbox.csproj]
Connected1: True
Connected2: False
 radekzikmund@radekzikmund-desktop  C:  source  DotnetSandbox                                                                                                                                                                                        10:07:33  

❯ wsl pwsh
PowerShell 7.2.3
Copyright (c) Microsoft Corporation.

Loading personal and system profiles took 1445ms.
 rzikm@radekzikmund-desktop    mnt  c  source  DotnetSandbox                                                                                                                                                                                       10:07:44  
❯ dotnet run
/mnt/c/source/DotnetSandbox/Program.cs(4,19): warning CS0618: 'TcpListener.TcpListener(int)' is obsolete: 'This constructor has been deprecated. Use TcpListener(IPAddress localaddr, int port) instead.' [/mnt/c/source/DotnetSandbox/DotnetSandbox.csproj]
Connected1: True
Connected2: True
```

File: src/libraries/System.Net.Sockets/src/System/Net/Sockets/Socket.cs [hunk 3]
```diff
@@ -700,6 +700,10 @@ public bool EnableBroadcast
         {
             get
             {
+                if (SocketType != SocketType.Dgram)
```
Comment (antonfirsov) [inline src/libraries/System.Net.Sockets/src/System/Net/Sockets/Socket.cs:703]:
```suggestion
                if (SocketType == SocketType.Stream)
```
Looks like `SO_BROADCAST` [also works for `SOCK_RAW`](https://docs.microsoft.com/en-us/windows/win32/winsock/tcp-ip-raw-sockets-2#send-and-receive-operations). Better be on the safe side and skip the OS PAL only for stream sockets.

File: src/libraries/System.Net.Sockets/tests/FunctionalTests/EnableBroadcastTest.cs [hunk 4]
```diff
@@ -0,0 +1,48 @@
...
+            using (Socket socket = new Socket(AddressFamily.InterNetwork, SocketType.Stream, ProtocolType.Tcp))
+            {
+                Assert.Throws<SocketException>(() =>
+                {
+                    socket.EnableBroadcast = true;
+                });
+            }
+        }
+
+
```
Comment (antonfirsov) [inline src/libraries/System.Net.Sockets/tests/FunctionalTests/EnableBroadcastTest.cs:32]:
```suggestion

```

extra newline

File: src/libraries/System.Net.Sockets/tests/FunctionalTests/EnableBroadcastTest.cs [hunk 5]
```diff
@@ -0,0 +1,46 @@
...
+        public void TcpConstructor_EnableBroadcast_SetterThrows()
+        {
+            using (Socket socket = new Socket(AddressFamily.InterNetwork, SocketType.Stream, ProtocolType.Tcp))
+            {
+                Assert.Throws<SocketException>(() =>
+                {
+                    socket.EnableBroadcast = true;
+                });
+            }
+        }
```
Comment (antonfirsov) [inline src/libraries/System.Net.Sockets/tests/FunctionalTests/EnableBroadcastTest.cs:30]:
```suggestion
        }

```

**Comment by turnyanszki** [comment]
Is there any problem with the code or thats just usual?

**Review by wfurt — APPROVED** [review]
LGTM

**Comment by wfurt** [comment]
outerloop tests tens to be noisy. I did not see anything related to this PR

**Review by antonfirsov — APPROVED** [review]

**Comment by antonfirsov** [comment]
@turnyanszki I got distracted by other topics, and forgot about this one, sorry. Merged now, thanks again for your contribution!

**Pull request merged**